libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -323,23 +544,246 @@ static MQTTStatus_t addRecord( MQTTPubAckInfo_t * records,
+                                    MQTTPublishState_t * pNewState )
+{
+    MQTTStatus_t status = MQTTIllegalState;
+    bool shouldDeleteRecord = false;
+    bool isTransitionValid = false;
+    MQTTPublishState_t newState = MQTTStateNull;
+
+    assert( records != NULL );
+    assert( pNewState != NULL );
+
+    newState = *pNewState;
+    shouldDeleteRecord = ( newState == MQTTPublishDone ) ? true : false;
+    isTransitionValid = validateTransitionAck( currentState, newState );
+
+    if( isTransitionValid == true )
+    {
+        /* Update record for acks. When sending or receiving acks for packets that
+         * are resent during a session reestablishment, the new state and
+         * current state can be the same. No update of record required in that case. */
+        if( currentState != newState )
+        {
+            updateRecord( records,
+                          recordIndex,
+                          newState,
+                          shouldDeleteRecord );
+        }
+
+        status = MQTTSuccess;
+    }
+    else
+    {
+        /* Invalid state transition. Update the new state to #MQTTStateNull. */
+        *pNewState = MQTTStateNull;
+        LogError( ( "Invalid transition from state %s to state %s.",
+                    MQTT_State_strerror( currentState ),
+                    MQTT_State_strerror( newState ) ) );
     }
+
+    return status;
 }
 
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t updateStatePublish( MQTTContext_t * pMqttContext,
+                                        size_t recordIndex,
+                                        uint16_t packetId,
+                                        MQTTStateOperation_t opType,
+                                        MQTTQoS_t qos,
+                                        MQTTPublishState_t currentState,
+                                        MQTTPublishState_t * pNewState )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t newState = MQTTStateNull;
+    bool isTransitionValid = false;
+
+    assert( pMqttContext != NULL );
+    assert( packetId != MQTT_PACKET_ID_INVALID );
+    assert( qos != MQTTQoS0 );
+    assert( pNewState != NULL );
+
+    newState = *pNewState;
+    isTransitionValid = validateTransitionPublish( currentState, newState, opType, qos );
+
+    if( isTransitionValid == true )
+    {
+        /* addRecord will check for collisions. */
+        if( opType == MQTT_RECEIVE )
+        {
+            status = addRecord( pMqttContext->incomingPublishRecords,
+                                MQTT_STATE_ARRAY_MAX_COUNT,
+                                packetId,
+                                qos,
+                                newState );
+        }
+        /* Send operation. */
+        else
+        {
+            /* Skip updating record when publish is resend and no state
+             * update is required. */
+            if( currentState != newState )
```

Comment:
Is this necessary? Even if the function is called it will only update to itself

Reply:
Why to do that if we know nothing to be done in this case?